HOSTEDCP-1707: Update version support and upgrade docs (#4347)
openshift/hypershift: docs/content/reference/versioning-support.md

# Versioning Support
There are different components that might require independent versioning and support level:

- Management Cluster
- API `hypershift.openshift.io`

> - HyperShift Operator (HO)

Reviewer: HCP is an "add-on" to MCE. I.e., HO should be replaced by Multi-cluster engine for d/s docs. HO is not versioned independently in the product and MCE is part of ACM.

- Control Plane Operator (CPO)
- HyperShift CLI
- Hosted Control Plane (HCP) CLI

## Support Level

> ### Management Cluster

Author: Do we have a minimum vanilla k8s version with respect to the HO or HC?

> There is no hard requirement on what OCP version the OpenShift management cluster needs to be running.

Reviewer 2: Something like this would help my understanding (note that I'm guessing at some of it).
```suggestion
The upstream HyperShift project does not place strict requirements on the OpenShift version of your management cluster. However, different versions of the HO are thoroughly tested only on a limited set of OpenShift versions, and this should inform your deployment decisions.

For a given build of the upstream HO (e.g. [v0.1.39](https://github.com/openshift/hypershift/releases/tag/v0.1.39)), the list of OCP versions tested for Management Cluster compatibility can be found here: `link to file in repository with this information -- ideally where it is programmatically checked and results in warnings if it is not met at runtime`.

For production use & support, it is required to use a downstream product which bundles a supported build of the HyperShift Operator. This downstream product is called [Multi-Cluster Engine](https://docs.openshift.com/container-platform/4.16/architecture/mce-overview-ocp.html) (MCE) and it is available through OpenShift's OperatorHub. 

MCE versions _do_ require specific OCP versions for the Management Cluster in order to remain in a supported. 
Each version documents its own support matrix. For example, 
- MCE 2.5: https://access.redhat.com/articles/7056007 . 
- MCE 2.4: https://access.redhat.com/articles/7027079 . 

As a heuristic, a new release of MCE will run on:
- The latest, yet to be released version of OpenShift.
- The latest GA version of OpenShift.
- 2 versions prior to the latest GA version.

Versions of MCE can also be obtained with the Advanced Cluster Management (ACM) offering. If you are running ACM, refer to product documentation to determine the bundled MCE version.
```

> The OCP versions of HostedClusters that can be installed on the management cluster depend on the HO version installed.

Reviewer 2: ```suggestion
The full list of HostedCluster OCP versions that can be installed via the HO on a Management Cluster will depend on the version of the installed HO. However, if you are running a tested configuration or MCE, this list will always include at least (a) the same OCP version as the Management Cluster and (b) Two previous minor versions relative to the Management Cluster. For example, if the Management Cluster is running 4.16 and a supported version of MCE, then the HO will at least be able to install 4.16, 4.15, and 4.14 Hosted Clusters.
```

### API
There are two user facing resources exposed by HyperShift: [HostedClusters and NodePools](https://hypershift-docs.netlify.app/reference/api/).

The HyperShift API version policy generally aligns with the [Kubernetes API versioning](https://kubernetes.io/docs/reference/using-api/#api-versioning).

### HO

> The HO is not tied to a particular OCP release. It is tied to the `main` branch of the HyperShift repo regardless of CPO version.

Reviewer 2: ```suggestion
The upstream HperShift project does not release new versions aligned with the OpenShift release cadence. New versions of the HO are periodically tagged from the `main` branch. These versions are tested and consumed by internal Red Hat managed services and you can use these versions directly. However, for supported, production use, you should use a supported version of MCE.
```

The HO is tagged at particular commits as part of merging new HO versions for Red Hat managed services; there is no particular tagging scheme for this effort.

A list of the tags can be found [here](https://github.com/openshift/hypershift/tags).

Once installed, the HO creates a ConfigMap called `supported-versions` into the Hypershift namespace, which describes the HostedClusters supported versions that could be deployed. 

Here is an example `supported-versions` ConfigMap:
```
apiVersion: v1
data:
    server-version: 2f6cfe21a0861dea3130f3bed0d3ae5553b8c28b
    supported-versions: '{"versions":["4.17","4.16","4.15","4.14"]}'
kind: ConfigMap
metadata:
    creationTimestamp: "2024-06-20T07:12:31Z"
    labels:
        hypershift.openshift.io/supported-versions: "true"
    name: supported-versions
    namespace: hypershift
    resourceVersion: "927029"
    uid: f6336f91-33d3-472d-b747-94abae725f70
```

!!! important

>         You cannot install HCs higher than what the HO supports. In the example above, HCs using 4.17 images cannot be created.

Reviewer 3: Should this say images > 4.17

Author: Yes 🤦🏻 - I updated the example configmap after I wrote that.

### CPO

> The CPO is released as part of each OCP payload release image. You can find those release images here:

Reviewer 2: Documentation on the CPO and how it supports or relates to the support matrix?

Author: Not sure which support matrix you're referencing? FWIW there is no direct support matrix between hub (management) OCP and OCP in hosted clusters.

# Versioning Support
There are different components that might require independent versioning and support level:

- Management Cluster
- API `hypershift.openshift.io`
- HyperShift Operator (HO)
- Control Plane Operator (CPO)
- HyperShift CLI
- Hosted Control Plane (HCP) CLI

## Support Level
### Management Cluster

> The upstream HyperShift project does not place strict requirements on the OpenShift version of your management cluster. The HO determines what versions of OCP can be installed through the HostedCluster (HC); see the [HO section](#ho) for more details. However, different versions of the HO are thoroughly tested only on a limited set of OpenShift versions, and this should inform your deployment decisions.

Reviewer 4: > The upstream HyperShift project does not place strict requirements on the OpenShift version of your management cluster

This is not entirely true. There are API versions that we use in our kube clients that require a particular version of Kube/OCP. There's also providers like Kubevirt which require 4.14 OCP.

> For a given build of the upstream HO (e.g. [v0.1.39](https://github.com/openshift/hypershift/releases/tag/v0.1.39)), the list of OCP versions tested for Management Cluster compatibility can be found here: `link to file in repository with this information -- ideally where it is programmatically checked and results in warnings if it is not met at runtime`.

Author: We may need to delete this line. I don't think we have this information readily available 🤔

Author: I'm deleting this line for now since we don't seem to have this info.

MCE versions _do_ require specific OCP versions for the Management Cluster to remain in a supported state. 
Each version documents its own support matrix. For example, 

- [MCE 2.5](https://access.redhat.com/articles/7056007)
- [MCE 2.4](https://access.redhat.com/articles/7027079)

As a heuristic, a new release of MCE will run on:

- The latest, yet to be released version of OpenShift
- The latest GA version of OpenShift
- Two versions prior to the latest GA version

Versions of MCE can also be obtained with the Advanced Cluster Management (ACM) offering. If you are running ACM, refer to product documentation to determine the bundled MCE version.

> The full list of HostedCluster OCP versions that can be installed via the HO on a Management Cluster will depend on the version of the installed HO. However, if you are running a tested configuration or MCE, this list will always include at least (a) the same OCP version as the Management Cluster and (b) Two previous minor versions relative to the Management Cluster. For example, if the Management Cluster is running 4.16 and a supported version of MCE, then the HO will at least be able to install 4.16, 4.15, and 4.14 Hosted Clusters. See the Multi-Cluster Engine section, under the expanded section titled "OpenShift Advanced Cluster Management" on [this page](https://access.redhat.com/support/policy/updates/openshift_operators) for more details. 

Reviewer 4: > (a) the same OCP version as the Management Cluster

If we are talking about downstream, that is true. But upstream hypershift project from main branch (4.17) can be installed on a 4.14 OCP mgmt cluster and support versions 4.17,4.16,4.15, and 4.14.

    namespace: hypershift
    resourceVersion: "927029"
    uid: f6336f91-33d3-472d-b747-94abae725f70
```

!!! important

        You cannot install HCs higher than what the HO supports. In the example above, HCs using images greater than 4.17 cannot be created.

### CPO
The CPO is released as part of each OCP payload release image. You can find those release images here:

- [amd64](https://amd64.ocp.releases.ci.openshift.org/)
- [arm64](https://arm64.ocp.releases.ci.openshift.org/)
- [multi-arch](https://multi.ocp.releases.ci.openshift.org/)

### HyperShift CLI
The HyperShift CLI is a helper utility used only for development and testing purposes. No compatibility policies are guaranteed.

It helps create required infrastructure needed for a HostedCluster CR and NodePool CR to successfully install. 

Running the following command will show what the latest OCP version the CLI supports against your KUBECONFIG:
```
% ./bin/hypershift version
Client Version: openshift/hypershift: 1ed535a8d27c5a1546f1ff4cc71abf32dd1a26aa. Latest supported OCP: 4.17.0
Server Version: 2f6cfe21a0861dea3130f3bed0d3ae5553b8c28b
Server Supports OCP Versions: 4.17, 4.16, 4.15, 4.14
```

### HCP CLI

> The HCP CLI is the productized version of the HyperShift CLI. This CLI is available through download through MCE. 

Reviewer 4: s/through MCE/from MCE/

For production use & support, it is required to use a downstream product which bundles a supported build of the HyperShift Operator. This downstream product is called [Multi-Cluster Engine](https://docs.openshift.com/container-platform/4.16/architecture/mce-overview-ocp.html) (MCE) and it is available through OpenShift's OperatorHub. 

MCE versions _do_ require specific OCP versions for the Management Cluster to remain in a supported state. 
Each version documents its own support matrix. For example, 

- [MCE 2.5](https://access.redhat.com/articles/7056007)
- [MCE 2.4](https://access.redhat.com/articles/7027079)

As a heuristic, a new release of MCE will run on:

- The latest, yet to be released version of OpenShift
- The latest GA version of OpenShift
- Two versions prior to the latest GA version

Versions of MCE can also be obtained with the Advanced Cluster Management (ACM) offering. If you are running ACM, refer to product documentation to determine the bundled MCE version.

> The full list of HostedCluster OCP versions that can be installed via the HO on a Management Cluster will depend on the version of the installed HO. However, if you are running a tested configuration or MCE, this list will always include at least (a) the same OCP version as the Management Cluster and (b) Two previous minor versions relative to the Management Cluster. For example, if the Management Cluster is running 4.16 and a supported version of MCE, then the HO will at least be able to install 4.16, 4.15, and 4.14 Hosted Clusters.

Reviewer: we can add a link to OCP lifecycle (MCE and ACM are documented there): 
![image](https://github.com/user-attachments/assets/4839f2e4-95d1-419c-9e43-175a514781dd)
 https://access.redhat.com/support/policy/updates/openshift_operators

Author: Added it.

> The full list of HostedCluster OCP versions that can be installed via the HO on a Management Cluster will depend on the version of the installed HO. However, if you are running a tested configuration or MCE, this list will always include at least (a) the same OCP version as the Management Cluster and (b) Two previous minor versions relative to the Management Cluster. For example, if the Management Cluster is running 4.16 and a supported version of MCE, then the HO will at least be able to install 4.16, 4.15, and 4.14 Hosted Clusters.

Reviewer 5: Another thing that we can clarify here is that the management cluster's OCP version itself does not restrict hosted clusters' OCP version that HO can install or upgrade. There is no relationship or interaction between the management cluster's OCP version and hosted cluster's OCP version.
